### ISSUE TITLE
Spurious warning about dangerous reinstalls when we're in a sandbox

### ISSUE THREAD

**Issue by 23Skidoo** [haskell/cabal#2101]
Sometimes one can get a warning about dangerous reinstalls even when one is using a freshly initialised sandbox. In my case Cabal needed to install `process-1.2.0.0`, which produced a warning because I already had `process-1.2.0.0` in global DB. Workaround: `--force-reinstalls`.

Example (trying to compile 0d83e52cc66d2f0eca8e2ef2e9821a3beb790d19 against time-1.5):

```
$ cabal install --only-dep --constraint "time >= 1.5"          
Resolving dependencies...
In order, the following would be installed:
stm-2.4.3 (new package)
text-1.2.0.0 (new package)
time-1.5 (new version)
random-1.0.1.3 (new package)
transformers-0.4.1.0 (new version)
mtl-2.2.1 (new package)
parsec-3.1.6 (new package)
network-uri-2.6.0.1 (new package)
unix-2.7.0.2 (new version)
directory-1.2.1.1 (new version)
network-2.6.0.2 (new package)
HTTP-4000.2.18 (new package)
process-1.2.0.0 (reinstall) changes: directory-1.2.1.0 -> 1.2.1.1,
unix-2.7.0.1 -> 2.7.0.2
Cabal-1.21.1.0 (new version)
zlib-0.5.4.1 (new package)
cabal: The following packages are likely to be broken by the reinstalls:
haskell98-2.0.0.3
ghc-7.8.3
Cabal-1.18.1.3
bin-package-db-0.0.0.0
Use --force-reinstalls if you want to install anyway.
```

**Comment by ygale** [comment]
I vote for this issue. Building in a sandbox to use dependency versions that are different than what's in the global and user package dbs is not only not dangerous and won't break anything - it is best practice. A scary warning is a serious usability bug here, in my opinion.

**Comment by kosmikus** [comment]
Yes, I think I also agree. An explicit `--force-reinstalls` should probably only be necessary if packages in the sandbox would be broken (and I'm not sure this is ever supposed to happen), but not if packages from the global DB would be broken.

**Comment by ygale** [comment]
Ah, I see what you mean. Is the following scenario possible? A sandbox build plan installs locally a new version of a package A that exists in the global DB. Another package B that exists in the global DB is also a dependency. The version of B in the global DB is compatible with both versions of A. Cabal decides that the installed version of B in the global DB is fine, and does not rebuild it in the sandbox.  The result is a broken build, since B was built against the global DB version of A.

If that can happen, cabal should be taught to detect that situation and re-install B in the sandbox. 

Ideally we should also have the solver drop the preference for the installed version of B, since anyway B will be re-installed, but it's not obvious what the algorithm for that should be.

**Comment by ygale** [comment]
> it's not obvious what the algorithm for that should be.

I thought of an algorithm, but I won't even write it here, because I'm afraid it will make it harder to get this issue implemented. Unless someone can think of a simple algorithm that will actually get done, I propose that we leave the solver alone for now and open a separate issue for that. Let's just make sure that the scenario I described above won't happen, and remove the warning for packages in a static DB that get rebuilt in a sandbox.

**Comment by kosmikus** [comment]
@ygale No, this scenario cannot occur. If `B` is also a dependency, then it's considered by the solver and the single-instance restriction forces a single version of `A`. So if `A` gets reinstalled, `B` has to be reinstalled as well.

**Comment by juhp** [comment]
This also happens for user installs vs global packages (ie outside a sandbox).

I currently have no user packages installed only global system packages
but I still get the error reported here (btw it is an error not just a warning:)
eg if I try to "cabal install stack" say in Fedora 22 with the Fedora Haskell
libraries installed.

I assume that is basically the same problem and that this is not
specific to sandboxing.

**Comment by ekenny** [comment]
I'm seeing a similar issue.  I started with a clean Ubuntu 15.10 installation, ran "sudo apt-get install haskell-platform" and then installed the eclipseFp plugin, which installed a bunch of packages (not sure if that matters.)  

Then I cloned Cabal and ran setup-dev.sh, which failed with this same message.  I'm attaching the full output of my latest run.  

Someone suggested on IRC that it should be safe to just run with --force-reinstalls, so I switched to the cabal-install directory and executed the same cabal command that setup-dev.sh uses (and added that switch), but it failed with a dependency-hell message.

[setup-dev.log.txt](https://github.com/haskell/cabal/files/162422/setup-dev.log.txt)

**Comment by ekenny** [comment]
I had forgotten that there was a failure prior to the failure I mentioned above.  The unit tests were failing, so I commented out the line in setup-dev.sh that runs the tests just to see if I could get a build.  That's the point at which I started seeing the issue above.  

I'm attaching the output of setup-dev.sh.  This happens with a completely clean machine.  I only installed haskell-package and git, cloned the Cabal repo and ran setup-dev.sh.

[setup-dev-sh-log-tests-enabled.txt](https://github.com/haskell/cabal/files/165478/setup-dev-sh-log-tests-enabled.txt)
